valuable-serde/src/lib.rs

```diff
@@ -0,0 +1,485 @@
+//! [`serde::Serialize`] implementation for [`valuable::Value`].
+//!
+//! # Examples
+//!
+//! ```
+//! use valuable::Valuable;
+//! use valuable_serde::SerdeValue;
+//!
+//! #[derive(Valuable)]
+//! struct Point {
+//!     x: i32,
+//!     y: i32,
+//! }
+//!
+//! let point = Point { x: 1, y: 2 };
+//!
+//! let value = SerdeValue::new(&point);
+//!
+//! assert_eq!(
+//!     serde_json::to_string(&value).unwrap(),
+//!     r#"{"x":1,"y":2}"#,
+//! );
+//! ```
+
+#![no_std]
+#![warn(missing_docs, rust_2018_idioms)]
+
+use core::mem;
+
+use serde::ser::{
+    Error, SerializeMap, SerializeSeq, SerializeStruct, SerializeStructVariant,
+    SerializeTupleStruct, SerializeTupleVariant,
+};
+use serde::{Serialize, Serializer};
+use valuable::field::Fields;
+use valuable::{EnumDef, NamedValues, StructDef, Valuable, Value, Variant, VariantDef, Visit};
+
+/// A wrapper around [`Value`] that implements [`Serialize`].
+#[derive(Debug, Clone, Copy)]
+pub struct SerdeValue<'a>(Value<'a>);
+
+impl<'a> SerdeValue<'a> {
+    /// Creates a new `SerdeValue`.
+    pub fn new<V>(v: &'a V) -> Self
+    where
+        V: Valuable,
+    {
+        Self(v.as_value())
+    }
+}
+
+impl<'a> From<Value<'a>> for SerdeValue<'a> {
+    /// Creates a new `SerdeValue` from [`Value`].
+    fn from(v: Value<'a>) -> Self {
+        Self(v)
+    }
+}
```

Comment:
An alternative idea is to store a type that implements `Valuable`.

```rust
pub struct SerdeValue<V>(V);

impl<V> SerdeValue<V>
where
    V: Valuable,
{
    pub fn new(v: V) -> Self {
        Self(v)
    }
}
```

Reply:
Making it generic seems good because then we don't have to worry about the lifetime.

Also, since this only supports serializing, maybe it should be called `struct Serializable<T>(T)` :laughing: going all-in w/ the *able pattern.

Reply:
Renamed to Serializable and made it generic.